src/kudu/consensus/routing.cc

```diff
@@ -201,6 +202,18 @@ void RoutingTable::ConstructNextHopIndicesRec(Node* cur) {
 Status RoutingTable::NextHop(const string& src_uuid,
                              const string& dest_uuid,
                              string* next_hop) const {
+  // Base case: use direct routing if no routing topology is defined. If we
```

Comment:
Just for my understanding - what makes this node assume that it has to go through the leader again to reach destination? Does this mean there can be multiple hops in a proxy. My limited understanding of how proxying works is: leader sends two requests to another node - one is for the node itself and another is a 'proxy' request for which this node builds a DownstreamRequest with the right payload and send it to destination node - the 'destination' being populated and explicitly indicated by the request sent by the leader. But the explanation below makes me believe that there could be multiple proxy hops.

Reply:
Hi Vinay, thanks for taking a look!

You're correct that the leader sends individual messages -- one with a payload to the next-hop node, and another message without any payload to the downstream node (and proxied through the next-hop node). You're also correct that multiple proxy hops are allowed, however we don't plan to use them in production immediately for UDB.

The model used for proxy routing (see comments at the top of routing.h) is a spanning tree, where the tree is defined as follows:

1. Explicit undirected edges are defined in the proxy topology config, which is a list of (`src`, `dest`) edges
2. Implicit undirected edges are generated from the leader to all remaining replicas that are not listed as a `dest` in any of the explicit edges in the proxy topology config (i.e. there is a direct path from the leader to any node that doesn't have a source proxy defined for it)

Once that spanning tree is constructed, there is only one way for a message to pass from any node `A` to another node `B`, and that is by following the edges in the tree. That means that inconsistency of proxy topology config between nodes can create a routing loop. If the leader `L` sends a node `P` a message intended to be proxied downstream to node `D`, but `P` has an empty proxy topology config, then without this patch the only way that `P` knows to get to `D` is to route the message back through `L`, since from `P`'s perspective the spanning tree  looks like this:

```
  L
 / \
P   D
```

And that makes sense because the default topology for Raft is a "star" topology.

This patch makes it so that we don't rely on the normal routing rules when handling a message to be proxied on a node with an empty proxy topology config. Instead, we will directly pass the message on to the destination node without trying to find a proper route, which makes more sense than a routing loop.